aries_vcx/src/new_protocols/mod.rs

```diff
@@ -0,0 +1,425 @@
+            "faber".to_owned(),
+            vec![faber_invite_verkey.clone()],
+            vec![],
+            dummy_endpoint.clone(),
+        );
+        let invitation = Invitation::with_decorators(
+            Uuid::new_v4().to_string(),
+            InvitationContent::Pairwise(invitation_content),
+            Default::default(),
+        );
+
+        let request_id = Uuid::new_v4().to_string();
+
+        let thread = match &invitation.content {
+            InvitationContent::Public(_) => {
+                let mut thread = Thread::new(request_id.clone());
+                thread.pthid = Some(invitation.id.clone());
+                thread
+            }
+            InvitationContent::Pairwise(_) => Thread::new(invitation.id.clone()),
+            InvitationContent::PairwiseDID(_) => Thread::new(invitation.id.clone()),
+        };
+
+        // Extract info from invitation
+        let bootstrap_info = alice.process_connection_invitation(invitation.content).await?;
+
+        // Build request
+        let (alice_did, alice_verkey) = alice.wallet().create_and_store_my_did(None, None).await?;
+        let mut did_doc = AriesDidDoc {
+            id: alice_did.clone(),
+            ..Default::default()
+        };
+        did_doc.set_service_endpoint(bootstrap_info.service_endpoint);
+        did_doc.set_routing_keys(bootstrap_info.routing_keys);
+        did_doc.set_recipient_keys(bootstrap_info.recipient_keys);
+
+        let con_data = ConnectionData::new(alice_did, did_doc);
+
+        let request_content = RequestContent::new("my_request".to_owned(), con_data);
+        let timing = Timing {
+            out_time: Some(Utc::now()),
+            ..Default::default()
+        };
+
+        let request_decorators = RequestDecorators {
+            thread: Some(thread),
+            timing: Some(timing),
+        };
+
+        let request = Request::with_decorators(request_id, request_content, request_decorators);
+
+        // Process request
+        let alice_did_doc = faber.process_connection_request(request.content).await?;
+        let recipient_keys = alice_did_doc.recipient_keys()?;
+        let their_verkey = recipient_keys.first().ok_or("no recipient keys")?.as_str();
+
+        // Build response
+        let (faber_did, faber_verkey) = faber.wallet().create_and_store_my_did(None, None).await?;
+        let response_content = faber
+            .build_response_content(their_verkey, faber_did, vec![faber_verkey], dummy_endpoint, vec![])
+            .await?;
+
+        let timing = Timing {
+            out_time: Some(Utc::now()),
+            ..Default::default()
+        };
+
+        let response_decorators = ResponseDecorators {
+            thread: Thread::new(request.decorators.thread.map(|t| t.thid).unwrap_or(request.id)),
+            please_ack: None,
+            timing: Some(timing),
+        };
+
+        let response = Response::with_decorators(Uuid::new_v4().to_string(), response_content, response_decorators);
```

Comment:
This lower level approach - removing state machines - is exposing the lack and the need for message builder.

Reply:
It does, yes.